client.go

```diff
@@ -976,16 +976,26 @@ func (c *Client) RemoveAll(path string) error {
 type File struct {
 	c      *Client
 	path   string
-	handle string
 
-	mu     sync.Mutex
+	mu     sync.RWMutex
+	handle string
 	offset int64 // current offset within remote file
 }
 
 // Close closes the File, rendering it unusable for I/O. It returns an
 // error, if any.
 func (f *File) Close() error {
-	return f.c.close(f.handle)
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if f.handle == "" {
+		return os.ErrClosed
+	}
+
+	handle := f.handle
+	f.handle = ""
```

Comment:
This is the meat of the fix for double closes.

Though I think we might want to change this to:
```
err := f.c.close(f.handle)
if err == nil {
	f.handle = ""
}
```

Actually… no. Looking up the code, `openssh/sftp-server.c` will release the handle no matter what the result of the `close` operation is.

So this is correct. (TODO: add a comment explaining this.)